### ISSUE TITLE
Does `SemiFuture` require `DefaultConstructible`?

### ISSUE THREAD

**Issue by brycelelbach** [executors/futures#19]
At the face to face meeting at Facebook we established that `SemiFuture`s must be `Movable`. What about `DefaultConstructible`?

**Comment by dhollman** [comment]
(Noting connection to issue #16 from the first call for future reference)

**Comment by brycelelbach** [comment]
@dhollman and @jaredhoberock please review.

**Comment by jaredhoberock** [comment]
I don't think we should require `DefaultConstructible` because it seems like a high bar for all futures to cross. For example, some futures might contain reference members, and it's difficult to create a reference out of thin air. And for a concrete example, `always_ready_future<T>` just can't be `DefaultConstructible` unless `T` is void. Its constructor needs to take a ready value, and not all value types are `DefaultConstructible`.

**Comment by LeeHowes** [comment]
For the concept I don't think we should require it be constructible at all. It is implementation-defined (ie executor-defined) how the future is obtained.

**Comment by dhollman** [comment]
> For the concept I don't think we should require it be constructible at all.

It's hard to write generic code without at least a move constructor. But yes, I agree that there is no reason to require `DefaultConstructible`

**Comment by LeeHowes** [comment]
Sorry, move is fine, yes. Constructible from anything other than itself is what I meant.
